File: regulations.py
import json
import os
import math

class RegulationsLoader:
    def __init__(self, json_path=None):
        if json_path is None:
            # Default to regulations.json in the same directory
            base_dir = os.path.dirname(os.path.abspath(__file__))
            json_path = os.path.join(base_dir, "regulations.json")
        
        self.json_path = json_path
        self.data = self._load_data()
        self.table_11 = self.data.get("table_11_flow_params", {}).get("data", [])
        self.table_12 = self.data.get("table_12_narrow_doors", {}).get("data", [])
        self.limits = self.data.get("limits", {})

    def _load_data(self):
        try:
            with open(self.json_path, 'r', encoding='utf-8') as f:
                return json.load(f)
        except Exception as e:
            print(f"Error loading regulations.json: {e}")
            return {}

    def get_flow_params(self, density, path_type):
        """
        Retrieves v (speed) and q (specific throughput) from Table 11.
        Uses stricter 'Next Higher Value' interpolation for density.
        
        path_type: "horizontal", "stair_down", "stair_up", "door"
        """
        # Clamp density to max 9.2 (boundary condition)
        density = min(density, 9.2)
        
        # Mapping path_type to json keys
        type_map = {
            "horizontal": "horiz",
            "stair_down": "stair_down",
            "stair_up": "stair_up",
            "door": "door_wide"
        }
        
        key = type_map.get(path_type)
        if not key:
            raise ValueError(f"Unknown path_type: {path_type}")

        # Find the row with D >= density (Next Higher Value)
        selected_row = None
        for row in self.table_11:
            if row["D"] >= density:
                selected_row = row
                break
        
        # If density is higher than max in table, use the last row (should coincide with clamp)
        if selected_row is None:
            selected_row = self.table_11[-1]

        params = selected_row[key]
        return params["v"], params["q"]

    def get_narrow_door_params(self, width):
        """
        Retrieves q and v for narrow doors (0.6 <= width <= 1.6) at boundary density (9.2).
        Uses linear interpolation.
        """
        # Clamp width to table range
        width = max(0.6, min(width, 1.6))
        
        # Find bracketing rows
        lower = None
        upper = None
        
        for row in self.table_12:
            if row["width"] == width:
                return row["v"], row["q"]
            if row["width"] < width:
                lower = row
            if row["width"] > width and upper is None:
                upper = row
                break
        
        if lower and upper:
            # Interpolate
            ratio = (width - lower["width"]) / (upper["width"] - lower["width"])
            v = lower["v"] + (upper["v"] - lower["v"]) * ratio
            q = lower["q"] + (upper["q"] - lower["q"]) * ratio
            return v, q
            
        # Fallback (shouldn't happen with clamping)
        entry = lower if lower else upper
        return entry["v"], entry["q"]

    def get_limit(self, path_type, param):
        """
        Get q_max or q_gran for a path type.
        path_type: "horizontal", "stairs_down", "stairs_up", "doors"
        param: "q_max", "q_gran", "v_gran"
        """
        return self.limits.get(path_type, {}).get(param, 0.0)

    def get_permissible_time(self, category, sub_category=None):
        """
        Simple lookup for permissible time.
        In a real component, this might need expanded logic for all building types.
        """
        pt = self.data.get("permissible_time_limits", {})
        cat_data = pt.get(category)
        
        if isinstance(cat_data, dict):
            if sub_category:
                return cat_data.get(sub_category, 999.0) # 999 as 'unlimited' default logic placeholder
            else:
                 return cat_data
        elif isinstance(cat_data, (int, float)):
            return cat_data
            
        return 999.0

File: BG_Fire_Evac/src/Logic/Regulations.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Text.Json;

namespace BG_Fire_Evac.Logic
{
    public class Regulations
    {
        private JsonDocument _data;
        private List<JsonElement> _table11;
        private List<JsonElement> _table12;
        private JsonElement _limits;
        private JsonElement _permissibleTimeLimits;

        public Regulations()
        {
            LoadData();
        }

        private void LoadData()
        {
            try
            {
                var assembly = Assembly.GetExecutingAssembly();
                using (var stream = assembly.GetManifestResourceStream("BG_Fire_Evac.src.Resources.regulations.json"))
                using (var reader = new StreamReader(stream))
                {
                    string json = reader.ReadToEnd();
                    _data = JsonDocument.Parse(json);
                    
                    var root = _data.RootElement;
                    _table11 = root.GetProperty("table_11_flow_params").GetProperty("data").EnumerateArray().ToList();
                    _table12 = root.GetProperty("table_12_narrow_doors").GetProperty("data").EnumerateArray().ToList();
                    _limits = root.GetProperty("limits");
                    _permissibleTimeLimits = root.GetProperty("permissible_time_limits");
                }
            }
            catch (Exception ex)
            {
                // In Grasshopper, we might want to throw or log this
                throw new InvalidOperationException("Failed to load embedded regulations.json", ex);
            }
        }

        public (double v, double q) GetFlowParams(double density, string pathType)
        {
            // Clamp density to max 9.2
            density = Math.Min(density, 9.2);

            string key = pathType switch
            {
                "horizontal" => "horiz",
                "stair_down" => "stair_down",
                "stair_up" => "stair_up",
                "door" => "door_wide",
                _ => throw new ArgumentException($"Unknown path_type: {pathType}")
            };

            JsonElement? selectedRow = null;

            // Next Higher Value interpolation
            foreach (var row in _table11)
            {
                if (row.GetProperty("D").GetDouble() >= density)
                {
                    selectedRow = row;
                    break;
                }
            }

            if (selectedRow == null)
            {
                selectedRow = _table11.Last();
            }

            var paramsObj = selectedRow.Value.GetProperty(key);
            return (paramsObj.GetProperty("v").GetDouble(), paramsObj.GetProperty("q").GetDouble());
        }

        public double GetPermissibleTime(string category, string subCategory = null)
        {
            if (_permissibleTimeLimits.TryGetProperty(category, out JsonElement catData))
            {
                if (catData.ValueKind == JsonValueKind.Object && subCategory != null)
                {
                    if (catData.TryGetProperty(subCategory, out JsonElement subVal))
                    {
                        return subVal.GetDouble();
                    }
                }
                else if (catData.ValueKind == JsonValueKind.Number)
                {
                    return catData.GetDouble();
                }
            }
            return 999.0; // Default unlimited
        }
    }
}
